java/arcs/core/storage/StorageProxy.kt

```diff
@@ -63,6 +72,33 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperationAtTime, T>(
     private val store: StorageCommunicationEndpoint<Data, Op, T> =
         storeEndpointProvider.getStorageEndpoint(ProxyCallback(::onMessage))
 
+    private val outgoingMessagesChannel =
+        Channel<Pair<ProxyMessage<Data, Op, T>, CompletableDeferred<Boolean>>>(Channel.UNLIMITED)
+
+    init {
+        // Send messages generated by applyOp to the Store in the order in which they were received
+        // by:
+        //
+        // 1. Calling store.onProxyMessage() with the message for each message.
+        // 2. Buffering those calls so they happen in parallel/pipelined-with calling store.idle()
+        // 3. Once the store is self-identifying as idle, complete the deferred which was generated
+        //    at the same time as the message (and was returned to the caller)
+        outgoingMessagesChannel.consumeAsFlow()
+            .map { (message, deferredToComplete) ->
+                log.debug { "Sending operations to store" }
+                store.onProxyMessage(message)
+                log.debug { "Operations sent to store" }
+                deferredToComplete
+            }
+            .buffer()
+            .onEach {
+                withTimeoutOrNull(1000) { store.idle() }
```

Comment:
So if null is returned, the store never became idle, yet we will complete the deferrable, is this ok or could it lead to subtle bugs for someone await()ing downstream? Should we at least log when the timeout cancellation occurs?

Reply:
It will definitely lead to subtle bugs, but I just wanted to avoid a hang. I can add a log message if it times out. good idea. This whole chunk should be temporary until we change it around to coalescing jobs (or lazily joining) as well.